File: info/1. Алгоритмы/2. Сортировка/16. Быстрая сортировка. Разбиение Хоара/code sample/sampleHoar1.py
def quick_sort(sequince, lo_index=None, hi_index=None):
    if lo_index is None:
        lo_index = 0
    if hi_index is None:
        hi_index = len(sequince)-1
    if lo_index >= hi_index:
        return None
    h = partition(sequince, lo_index, hi_index)
    quick_sort(sequince, lo_index, h-1)
    quick_sort(sequince, h+1, hi_index)


def partition(sequince, lo_index, hi_index):
    support_element = sequince[lo_index]
    i = lo_index + 1
    j = hi_index
    while True:
        while i < hi_index and sequince[i] < support_element:
            i += 1
        while sequince[j] > support_element:
            j -= 1
        if i >= j:
            break
        sequince[i], sequince[j] = sequince[j], sequince[i]
        i += 1
        j -= 1
    sequince[lo_index], sequince[j] = sequince[j], sequince[lo_index]
    return j


sequince = [0, 5, -2, 7, 3]

quick_sort(sequince)

print(sequince)

File: info/1. Алгоритмы/2. Сортировка/16. Быстрая сортировка. Разбиение Хоара/code sample/SampleHoar1.java
package algoritms.sorts.quick_sort;

import java.util.Arrays;

public class SampleHoar1 {
	public static void main(String[] args) {
		int[] array = new int[] { 0, 5, -2, 7, 3, -2 };
		quickSort(array);
		System.out.println(Arrays.toString(array));

	}

	public static void quickSort(int[] array) {
		quickSort(array, 0, array.length - 1);
	}

	public static void quickSort(int[] array, int lo, int hi) {
		if (lo >= hi) {
			return;
		}
		int h = breakPartition(array, lo, hi);
		quickSort(array, lo, h - 1);
		quickSort(array, h + 1, hi);
	}

	public static int breakPartition(int[] array, int lo, int hi) {
		int i = lo + 1;
		int supportElement = array[lo];
		int j = hi;
		for (;;) {
			for (; i < hi && array[i] < supportElement;) {
				i += 1;
			}
			for (; array[j] > supportElement;) {
				j -= 1;
			}
			if (i >= j) {
				break;
			}
			swap(array, i++, j--);
		}
		swap(array, lo, j);
		return j;
	}

	public static void swap(int[] array, int i, int j) {
		int temp = array[i];
		array[i] = array[j];
		array[j] = temp;
	}
}
